zebra-state/src/service/finalized_state/zebra_db/block.rs

```diff
@@ -189,26 +193,69 @@ impl ZebraDb {
     ) -> Result<block::Hash, BoxError> {
         let finalized_hash = finalized.hash;
 
-        // Get a list of the spent UTXOs, before we delete any from the database
-        let all_utxos_spent_by_block: HashMap<transparent::OutPoint, transparent::Utxo> = finalized
-            .block
-            .transactions
+        let tx_hash_indexes: HashMap<transaction::Hash, usize> = finalized
+            .transaction_hashes
+            .iter()
+            .enumerate()
+            .map(|(index, hash)| (*hash, index))
+            .collect();
+
+        // Get a list of the new UTXOs in the format we need for database updates.
+        //
+        // TODO: index new_outputs by TransactionLocation,
+        //       simplify the spent_utxos location lookup code,
+        //       and remove the extra new_outputs_by_out_loc argument
+        let new_outputs_by_out_loc: BTreeMap<OutputLocation, transparent::Utxo> = finalized
+            .new_outputs
             .iter()
-            .flat_map(|tx| tx.inputs().iter())
-            .flat_map(|input| input.outpoint())
-            .map(|outpoint| {
+            .map(|(outpoint, utxo)| {
                 (
-                    outpoint,
-                    // Some utxos are spent in the same block, so they will be in `new_outputs`
-                    self.utxo(&outpoint)
-                        .or_else(|| finalized.new_outputs.get(&outpoint).cloned())
-                        .expect("already checked UTXO was in state or block"),
+                    lookup_out_loc(finalized.height, outpoint, &tx_hash_indexes),
+                    utxo.clone(),
                 )
             })
             .collect();
 
+        // Get a list of the spent UTXOs, before we delete any from the database
+        let spent_utxos: Vec<(transparent::OutPoint, OutputLocation, transparent::Utxo)> =
+            finalized
+                .block
+                .transactions
+                .iter()
+                .flat_map(|tx| tx.inputs().iter())
+                .flat_map(|input| input.outpoint())
+                .map(|outpoint| {
+                    (
+                        outpoint,
+                        // Some utxos are spent in the same block, so they will be in
+                        // `tx_hash_indexes` and `new_outputs`
+                        self.output_location(&outpoint).unwrap_or_else(|| {
+                            let tx_index = tx_hash_indexes
+                                .get(&outpoint.hash)
+                                .expect("already checked UTXO was in state or block");
+                            let tx_loc =
+                                TransactionLocation::from_usize(finalized.height, *tx_index);
+                            OutputLocation::from_outpoint(tx_loc, &outpoint)
```

Comment:
Minor: shouldn't this be a call to `lookup_out_loc`?

Reply:
Thanks, I must have missed that cleanup.

```suggestion
                            lookup_out_loc(finalized.height, outpoint, &tx_hash_indexes)
```